### PR TITLE
File api

### LINKED ISSUE #62: Retrieve build targets using CMake's file API
Vim-CMake uses the `help` target to retrieve a list of available build targets. When running `cmake --build <build_dir> --target help`, there are a few differences between the output for Makefile systems and that for Ninja systems—e.g. the format is different, and Ninja systems also return a lot more entries.

To make the retrieval of build targets consistent and build tool-independent, we could use [CMake's file api](https://cmake.org/cmake/help/latest/manual/cmake-file-api.7.html). From https://github.com/cdelledonne/vim-cmake/issues/57#issuecomment-1203294478:

> Fortunately CMake has a feature for this. It is called [cmake-file-api(7)](https://cmake.org/cmake/help/latest/manual/cmake-file-api.7.html) and is meant to be used by 'clients' to retrieve semantic information about the buildsystems CMake generates. So exactly what we want, especially with the [Object Kind "codemodel"](https://cmake.org/cmake/help/latest/manual/cmake-file-api.7.html#object-kind-codemodel).

Parsing JSON data should not be an issue, as both Vim and Neovim provide the function `json_decode()`.

### PR DESCRIPTION
**Pull request by marwing**
fixes #62

### CODE REVIEW

**Comment by cdelledonne** [comment]
Apologies for the long radio silence, I was on holiday and then again very busy with work. I'll review your PR in the next few days. A huge thanks in advance.

**Comment by marwing** [comment]
Great. This probably still has some rough edges and is missing docs. Just let me know what you find. I'm quite busy currently so it may take a while until I can address any issues.

**Review by cdelledonne — CHANGES_REQUESTED** [review]
Looking great! I left some mostly minor comments and questions, let me know if anything is unclear.

> This probably still has some rough edges and is missing docs.

What do you think we need to document? One thing I would mention in README and docs is that CMake 3.14 is required for build target autocompletion.

One more thing: please rebase onto the latest `master`, and see if we need to add anything to `s:buildsys.Init()`, which is called also when the user jumps to another project (perhaps we need to refresh targets there?)

File: autoload/cmake/buildsys.vim
Comment (marwing) [reply autoload/cmake/buildsys.vim:361]: Because switching config may change available targets depending on CMake configuration (e.g. a test target that only shows up in the Debug config).

File: autoload/cmake/buildsys.vim
Comment (marwing) [reply autoload/cmake/buildsys.vim:496]: Because the available targets may change from outside factors (e.g. user running CMake from the cli). `buildsys` doesn't have a local version of targets anymore and `s:fileapi.Reparse` is already checking if the index changed. I figured it would be easier and more correct to just let the fileapi handle update detection. If this turns out to be too much overhead this can be removed but since most of the time the cached version should be correct this should be OK.

File: autoload/cmake/build.vim
Comment (marwing) [reply autoload/cmake/build.vim:78]: Whoops. It's not. Missed this.

File: autoload/cmake/fileapi.vim
Comment (marwing) [reply autoload/cmake/fileapi.vim:2]: My bad.

File: autoload/cmake/fileapi.vim
Comment (marwing) [reply autoload/cmake/fileapi.vim:3]: My bad. Should have enabled spellchecking in comments.

File: autoload/cmake/fileapi.vim
Comment (marwing) [reply autoload/cmake/fileapi.vim:8]: We can. But this would duplicate the initial state between here and `Reset()`. Since the others do this too, I'll change it anyways.

File: autoload/cmake/fileapi.vim
Comment (marwing) [reply autoload/cmake/fileapi.vim:131]: Leftover from experimenting with query versioning. Will remove.

File: autoload/cmake/fileapi.vim
Comment (marwing) [reply autoload/cmake/fileapi.vim:135]: Having `index` here saves me from returning returning it from `ParseIndex` which already has a return value. It could be a script local variable instead of a class variable though. Otherwise you are correct. Having codemodel as a dict is just a thing I did because we might want to add more keys to it, but this would be easy enough to change in the event we did.

File: autoload/cmake/fileapi.vim
Comment (marwing) [reply autoload/cmake/fileapi.vim:132]: My bad. The others all are.

File: autoload/cmake/fileapi.vim
Comment (marwing) [reply autoload/cmake/fileapi.vim:156]: I don't know. I had it as `Parse` before but changed it. Both sort of work but are slightly inaccurate. If you think `Parse` is better I will change it.

File: autoload/cmake/fileapi.vim
Comment (marwing) [reply autoload/cmake/fileapi.vim:34]: I hadn't thought of that. Although it probably doesn't matter too much in this file since this information is either short lived or non crucial. Will change anyway.

File: autoload/cmake/fileapi.vim
Comment (marwing) [reply autoload/cmake/fileapi.vim:131]: Another thing I missed. Let's see if I ever get the hang of vimscript.

File: autoload/cmake/fileapi.vim
Comment (marwing) [reply autoload/cmake/fileapi.vim:135]: Missed the `codemodel_file` part. This is the only part used from the index outside the `ParseIndex` function. This once again is a "we may want to add more later". Will change. The entire logic to detect the codemodel path seems weird now. I don't even use the results...

**Comment by marwing** [comment]
I think I addressed all your comments by either a change or an explanation. I also rebased on master. I don't think we have to do anything special in `s:buildsys.Init()`. You already call `s:SetCurrentConfig` which does the necessary refreshing.

For documentation I think we should at least document the new version requirement and the fact that the file api is used at all. The rest is internals I would say. Also the changelog entry is still missing.

File: autoload/cmake/fileapi.vim
Comment (cdelledonne) [reply autoload/cmake/fileapi.vim:11]: We can skip calling `Reset()` at initialization time now, can't we?

File: autoload/cmake/fileapi.vim
Comment (cdelledonne) [reply autoload/cmake/fileapi.vim]: Not the most sensible language out there (:

File: autoload/cmake/fileapi.vim
Comment (cdelledonne) [reply autoload/cmake/fileapi.vim]: Actually the result of `s:FindIndexFile()` is not short lived, it gets saved into `s:fileapi.last_index_name`. If, after running this once, we `:cd` into a subfolder of the root folder, and obtain the new relative path to the index file, `s:ParseIndex()` will think the new file name differs from the previous one.

File: autoload/cmake/buildsys.vim
Comment (cdelledonne) [reply autoload/cmake/buildsys.vim:517]: I think the overhead would indeed be negligible

File: autoload/cmake/buildsys.vim
Comment (cdelledonne) [inline autoload/cmake/buildsys.vim:447]: Is this call to `Parse()` necessary? We will parse anyway in the `s:RefreshTargets()` callback, won't we?

File: autoload/cmake/buildsys.vim
Comment (cdelledonne) [reply autoload/cmake/buildsys.vim:362]: True, but we don't always need to refresh targets in this function. For instance, the first time we call `s:SetCurrentConfig()` is inside `s:ProcessBuildConfig()`, when we're processing the command-line arguments passed by the user with `:CMakeGenerate`. At that point, the build system does not exist yet, so conceptually there's not point in trying to refresh targets.

I would suggest to explicitly call `s:RefreshTargets()` where really needed (`s:buildsys.Init()` and `s:buildsys.Switch()`). This would also make the logic more clear for someone who reads the code in the future, in my opinion.

**Comment by cdelledonne** [comment]
Thanks a lot for the quick reaction, back to you with a couple more comments!

File: autoload/cmake/fileapi.vim
Comment (marwing) [reply autoload/cmake/fileapi.vim]: While the result is not short lived it is only used in a caching manner. Changing directory would essentially just invalidate the cache. Nothing would break because of this (most likely, CMake names its index file with a timestamp down to the millisecond). In any case, it is more correct to save the full path.

File: autoload/cmake/buildsys.vim
Comment (marwing) [reply autoload/cmake/buildsys.vim:447]: Indeed

File: autoload/cmake/buildsys.vim
Comment (marwing) [reply autoload/cmake/buildsys.vim:362]: Fair enough

**Review by cdelledonne — COMMENTED** [review]
The changes look good to me! Once you're done with documentation stuff, I think we're ready to merge.

**Comment by marwing** [comment]
I think this should be enough for documentation since this is more of an internal change. Let me know if you would reword anything. Otherwise I would say this is ready.

File: README.md
Comment (cdelledonne) [inline README.md:18]: Actually, the file API is only used for build targets. Also, I think this is too low-level of a detail to be mentioned in the feature overview. Perhaps we can revert this and mention the file API in the requirements section? (see comment below)

File: README.md
Comment (cdelledonne) [inline README.md:28]: E.g. "Target autocompletion uses the `cmake-file-api(7)`, which requires CMake 3.14 or later"

Comment (cdelledonne) [inline README.md:28]: By the way, we should add this line to doc/cmake.txt as well

File: autoload/cmake/fileapi.vim
Comment (cdelledonne) [inline autoload/cmake/fileapi.vim:59]: The error message is in my opinion too abstract for the user, most people who will get this error won't know what "Api response missing" means. Could we make this message a bit more meaningful to the user? Is this too vague? "List of build targets missing, run `:CMakeGenerate`"

**Review by cdelledonne — CHANGES_REQUESTED** [review]
A couple of rewording suggestions, let me know what you think

File: README.md
Comment (marwing) [reply README.md:28]: It's not autocompletion, is it?

File: README.md
Comment (marwing) [reply README.md:18]: Sure.

File: README.md
Comment (marwing) [reply README.md:28]: Oops. I missed this section is in docs as well. Will add.

File: autoload/cmake/fileapi.vim
Comment (marwing) [reply autoload/cmake/fileapi.vim:59]: I know this is not a good message. I just noticed this would be a problem and added something to not forget. I think 'List of build targets' is too vague. I, unfortunately, don't have a better idea myself. The only thing I can think of would be to explicitly name the `cmake-file-api(7)` here too. This is only slightly better, but at least you could google (or `:Man`) what this is.

File: README.md
Comment (marwing) [reply README.md:28]: Should I leave [this](https://github.com/cdelledonne/vim-cmake/pull/64/files#diff-b73fc732b2dcef182f1fed830367b06d323bb053305436c27b6c6321d82862b3R140-R141) in then?

Comment (cdelledonne) [reply README.md:28]: I think just "completion" is enough

Comment (cdelledonne) [reply README.md:28]: I would leave this in, yes (just add a blank line right after these two lines, forgot to point it out earlier)

File: autoload/cmake/fileapi.vim
Comment (cdelledonne) [reply autoload/cmake/fileapi.vim:59]: I would not get too technical, as all the user cares about is in fact the latter part of the sentence. I also don't have any great suggestions now, so I'll leave this up to you, and we can refine in the future if necessary

Comment (marwing) [reply autoload/cmake/fileapi.vim:59]: I like to think that people using vim to write software in a language that uses CMake for a buildsystem are at least somewhat interested in technicalities. (Could be wrong though)
As I think 'Api response' is in fact not useful at all I will change this to include the reference to the file api. If we ever find some better message we can change it. I will also explicitly mention that target completion will not work because of this.

File: autoload/cmake/fileapi.vim
Comment (cdelledonne) [inline autoload/cmake/fileapi.vim:62]: I realized this warning gets echoed whatever command you run when there isn't a build system yet. Even if you just run `:CMakeGenerate` without a build system being there yet. Which is not what we want I guess.

Ideally, this warning gets echoed only if the user tries to tab-complete targets before generating a build system, i.e. only when this function is triggered as a consequence of invoking `cmake#GetBuildTargets()` (from autoload/cmake.vim). We should not issue this warning if `s:fileapi.Parse()` is invoked from other code paths.

Would you agree? To circumvent this problem, we could only raise an exception from this function, and silently ignore such exception everywhere except for in `cmake#GetBuildTargets()`, where we do echo the warning. Or something else that has the same effect.

Sorry I'm raising this just now, I hadn't tested this properly.

**Comment by cdelledonne** [comment]
Also, given that `s:buildsys.GetTargets()` is only called once now (from `s:GetBuildArgs()` in build.vim), we could get rid of that, and always use `s:fileapi` to retrieve targets.

File: autoload/cmake/fileapi.vim
Comment (marwing) [reply autoload/cmake/fileapi.vim:169]: I mostly agree, I don't know what I was thinking. I would throw directly from `s:ParseIndex()` however, as this allows us to throw different exception for different errors. If we still name them in a consistent way we still can catch `/vim-cmake_fileapi_.*/` in function in which we just don't care. This also allows us to return the response files array locally instead of using a 'global' variable and an error code.

**Comment by cdelledonne** [comment]
The latest version look great! Could I just ask you to update the docstrings of these functions in fileapi.vim? Where applicable, they should include `Params`, `Returns`, and perhaps also `Raises` for those that can raise exceptions.
- `s:ParseIndex()`
- `s:ParseCodemodel()`
- `s:fileapi.Parse()`

After that, I'll just merge, if you agree.

**Comment by cdelledonne** [comment]
And by the way, thanks A LOT for taking the time to work on this feature and address my review comments, I appreciate that a lot

**Comment by marwing** [comment]
> and perhaps also `Raises` for those that can raise exceptions.

Does this mean for all those from which a vim-cmake exception can escape or only those with the actual throw statement (does it include `s:fileapi.Parse()` or only `s:ParseIndex()`).

> After that, I'll just merge, if you agree.

I do.

**Comment by cdelledonne** [comment]
I meant to also document `Raises` in `s:fileapi.Parse()`, given that it's a public function, so that if we are to use it in some other places in the future we know how to treat it. It'd be great if you could apply this last one change. Thanks!

**Comment by cdelledonne** [comment]
That was fast! I'm merging

**Comment by marwing** [comment]
> That was fast! I'm merging

I may have had this prepared in case this is what you wanted... :)

Anyways, thanks for merging.

**Pull request merged**
